src/configureRoutes.js

```diff
@@ -5,12 +5,96 @@ import warning from './utils/warning';
 import requireModuleDefault from './utils/requireModuleDefault';
 import CatalogContext from './components/CatalogContext';
 import PageContent from './components/Page/PageContent';
-import fetchMarkdown from './fetchMarkdown';
+import Page from './components/Page/Page';
+import fetchMarkdown, {fetchMarkdownNoCatch} from './fetchMarkdown';
+
+// Use heuristic to detect of the value is a valid react component
+// (a value which can be given as the first argument to React.createElement()).
+//
+// It works like this: if it's a constructor and the instance has a 'render'
+// function it's good. That handles both ES6 classes and components created
+// with createClass(). And because we're not using instanceof checks, we're
+// independent of the React library which the user is using (might be
+// different from the one used by Catalog itself).
+const isReactComponent = (X) => {
+  try {
+    const v = new X();
+    return v !== null && typeof v === 'object' && typeof v.render === 'function';
+  } catch (e) {
+    return false;
+  }
+};
 
 const pageComponent = (page) => {
+  // page.content can be almost anything (as long as it's truthy) and
+  // we'll do our best to render it.
   if (page.content) {
-    return (props) => <PageContent {...props} contentPromiseFn={() => Promise.resolve(page.content())} />;
+    const content = page.content;
+
+    const gaveUpPage = () =>
+      <Page>{`\`\`\`hint|warning\n# Catalog: I gave up.\n\n I don't know how to handle this page content: "${content}"\n\`\`\``}</Page>;
+
+    return (props) => {
+      // This function takes anything and returns a Promise which resolves
+      // to a React$Element. It's magic.
+      const coerce = (x) => {
+        // 1. React$Element
+        // This is the termination condition of this recursive function.
+        if (React.isValidElement(x)) {
+          return Promise.resolve(x);
+        }
+
+        // 2. React Component
+        // And this is the second termination condition.
+        if (isReactComponent(x)) {
+          return Promise.resolve(React.createElement(x));
+        }
+
+        // 3. Promise
+        if (typeof x.then === 'function') {
+          return x.then(coerce).catch(e => <Page>{`\`\`\`hint|warning\nPromise failed.\n${e}\n\`\`\``}</Page>);
+        }
+
+        // 4. Function
+        // Invoke the function and coerce its return value.
+        if (typeof x === 'function') {
+          return coerce(x());
```

Comment:
It could be a function which returns a promise. For example `() => import('…')`, you want to wrap the `import()` in a function so the import is not executed immediately.